ddtrace/internal/writer.py

```diff
@@ -271,67 +270,57 @@ def _log_error_status(self, response):
-    will randomly overwrite elements when it's over the max size.
-
-    This queue also exposes some statistics about its length, the number of items dropped, etc.
-    """
-
+class TraceQueue:
     def __init__(self, maxsize=0):
-        # Cannot use super() here because Queue in Python2 is old style class
-        Queue.__init__(self, maxsize)
-        # Number of item dropped (queue full)
-        self.dropped = 0
-        # Number of items accepted
+        self._lock = threading.Lock()
+        self._queue = []
+        self.maxsize = maxsize
+
         self.accepted = 0
-        # Cumulative length of accepted items
         self.accepted_lengths = 0
+        self.dropped = 0
+
+    class Empty(Exception):
+        pass
+
+    @property
+    def queue(self):
+        return self._queue
+
+    def qsize(self):
+        return len(self._queue)
 
     def put(self, item):
-        try:
-            # Cannot use super() here because Queue in Python2 is old style class
-            Queue.put(self, item, block=False)
-        except Full:
-            # If the queue is full, replace a random item. We need to make sure
-            # the queue is not emptied was emptied in the meantime, so we lock
-            # check qsize value.
-            with self.mutex:
-                qsize = self._qsize()
-                if qsize >= self.maxsize:
-                    idx = random.randrange(0, qsize)
-                    self.queue[idx] = item
-                    log.warning("Writer queue is full has more than %d traces, some traces will be lost", self.maxsize)
-                    self.dropped += 1
-                    self._update_stats(item)
-                    return
-            # The queue has been emptied, simply retry putting item
-            return self.put(item)
-        else:
-            with self.mutex:
-                self._update_stats(item)
-
-    def _update_stats(self, item):
-        # self.mutex needs to be locked to make sure we don't lose data when resetting
-        self.accepted += 1
-        if hasattr(item, "__len__"):
-            item_length = len(item)
-        else:
-            item_length = 1
-        self.accepted_lengths += item_length
+        with self._lock:
+            if len(self._queue) < self.maxsize:
+                self._queue.append(item)
+            else:
+                idx = _rand.rand64bits() % len(self._queue)
+                self.queue[idx] = item
+
+                log.warning("Writer queue is full has more than %d traces, dropping trace", self.maxsize)
+                self.dropped += 1
+
+            self.accepted += 1
+            self.accepted_lengths += len(item) if hasattr(item, "__len__") else 1
+
+    def get(self, block=False):
+        with self._lock:
+            try:
+                if len(self._queue) == 0:
+                    raise self.Empty
+                return self._queue
+            finally:
+                self._queue = []
+
+    def pop_stats(self):
```

Comment:
Why isn't that called `reset_stats` directly?

Reply:
Yeah, all of these strange looking changes are left in to maintain the API. But I just made the realization that we're in the `internal` module anyway - so let's axe them!